apps/api/app/utils/search_utils.py

```
@@ -192,3 +206,161 @@ async def fetch_with_firecrawl(url: str, use_stealth: bool = False) -> str:
+      2. crawl4ai   — good JS support, free, no API key required
+      3. httpx+BS4  — always available, static pages only
+
+    Each tier is independently cached in Redis (1h TTL).
+    Raises FetchError only if all three tiers fail.
+    """
+    errors: list[str] = []
+
+    # Tier 1: Firecrawl
+    try:
+        return await fetch_with_firecrawl(url)
+    except Exception as e:
```

Ruff (0.15.2)

[warning] Do not catch blind exception: `Exception`

(BLE001)